### PR TITLE
cache locking

### PR DESCRIPTION
**Pull request by LunicLynx**
solves for #55 

Needs a proper review for other scenarios.

### CODE REVIEW

File: src/DryIoc/Container.cs [hunk 1]
```diff
@@ -335,7 +363,18 @@ private object ResolveAndCacheFactoryDelegate(Type serviceType, IfUnresolved ifU
             // Additionally disable caching when no services registered, not to cache an empty collection wrapper or alike.
             CacheKeyedFactoryDelegateOrExpression(cacheRef, key, cacheEntry, entryKey, factoryDelegate);
 
+            // we now cached the factoryDelegate so we do not need to lock again
+            lock (_locksLock)
```
Comment (LunicLynx) [inline src/DryIoc/Container.cs]: here could be:
`if(__lockWasTaken)`

File: src/DryIoc/Container.cs [hunk 2]
```diff
@@ -276,6 +276,34 @@ private object ResolveAndCacheFactoryDelegate(Type serviceType, IfUnresolved ifU
 
             FactoryDelegate factoryDelegate;
 
+            // get the lock state and object for the service type
+            Tuple<bool, object> requireLockAndLockObj;
+            lock (_locksLock)
```
Comment (LunicLynx) [inline src/DryIoc/Container.cs]: Maybe your ImHashMap Could handle this with only locking if required ?

**Comment by dadhi** [comment]
We need to fix a mess with `UseInstance`, without touching the general registration hot-path if possible. Maybe create a separate registry for instances, maybe make it ephemeral by getting the instances directly from scopes (but keyed registrations and `ifAlreadyRegistered` strategy do not play well here). Explore other ideas as well..

**Comment by LunicLynx** [comment]
This is what i thought as well. This was just the first idea that came to mind to fix it. It is actually not one problem but two.
The big issue is the cloning of the registry with the Ref<> to the Cache, which is solved by the lock when putting it in the cache. I guess that is fine, because it only happens until every factory delegate was created.
The second issue is if multiple threads run into the resolve at the same time before the delegate is cached. This can create the delegate as often as there are cores in the cpu, or even more often, dunno. But it will go away after the first delegate is put into the cache.

I dunno what the performance impact of one single lock is. I would suggest to measure it. I don't see a real way here to make it thread safe and reduce the factory delegation creation to one.

**Pull request closed without merging**